[liauzhanyi] iP

matchastack opened the pull request:
# ZBot

This PR adds **more functionalities** to ~~Duke~~ ZBot. 

The bot can now keep track of tasks, and you may even specify due dates, or a time period just like a calendar 📆 event!

## User Guide

To use the app, follow the steps below:

1. Clone the [repository](https://github.com/liauzhanyi/ip)
2. Run the program using 
```bash
./gradlew build
./gradlew run
```

## Command Summary

These are the currently supported commands

- `todo [task]`
- `deadline [task] /by [dd/MM/yyyy HHmm]`
- `event [task] /from [dd/MM/yyyy HHmm] /to [dd/MM/yyyy HHmm]`
- `list`
- `delete [index]`
- `find`
- `mark [index]` 
- `unmark [index]`

More features _coming soon_!

- [ ] GUI
- [ ] Support for more date input formats


For future developers, the aim of the project is to be
> everything to someone

and not 

> something to everyone

Since there is no syntax highlight for bash, this is just here for the sake of adding colours since as programmers, we all love _colourful words_

```java
import java.utils.*;

public static void main(String[] args) {
    new ZBot(SAVE_PATH).run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

matchastack committed:
Update .gitignore to ignore compiled class files

matchastack committed:
Replace Duke class with ZBot class

The Duke class has been replaced with the ZBot class to match the new chatbot name and all traces of Duke are removed.

matchastack committed:
Refactor ZBot class to improve code structure and readability

matchastack committed:
feat: Add user input handling to ZBot

The program now echos commands entered by the user, and exits when the user types the command 'bye'.

matchastack committed:
feat: Add list command

Add the ability to store whatever text entered by the user and display them back to the user when requested by issuing the 'list' command.

matchastack committed:
chore: Refactor ZBot class and add Task class

Refactor the ZBot class to improve code structure and readability. Additionally, add a new Task class to handle storing and displaying tasks entered by the user.

matchastack committed:
feat: Improve task display format

Refactor the `toString()` method in the `Task` class to display tasks in a more informative format.

matchastack committed:
feat: Add mark task as done/undone functionality

matchastack committed:
chore: Add exceptions handling in mark task functionality

Handle exceptions that arises when user issue mark/unmark command without a task number or invalid task number.

matchastack committed:
feat: Add ToDo class

This commit adds a new `ToDo` class that extends the `Task` class. The `ToDo` class is responsible for storing and displaying tasks entered by the user. It overrides the `toString()` method to display tasks in a specific format.

matchastack committed:
chore: Update ZBot's addTask method to improve task display format

matchastack committed:
feat: Add Deadline class

This commit adds a new `Deadline` class that extends the `Task` class. The `Deadline` class is responsible for storing and displaying tasks with a deadline entered by the user. It overrides the `toString()` method to display tasks in a specific format.

matchastack committed:
feat: Add Event class

This commit adds a new `Event` class that extends the `Task` class. The `Event` class is responsible for storing and displaying tasks with a specific event period entered by the user. It overrides the `toString()` method to display tasks in a specific format.

matchastack committed:
chore: Add semi-automated test

matchastack committed:
Update test

Include more lines to input.txt to make test more comprehensive.

matchastack committed:
fix: ArrayIndexOutOfBoundsException caused by todo

Catch possible error due to user input todo command without description.

matchastack committed:
refactor: Update ZBot's addTask method to use ArrayList for tasks

The ZBot class previously used an array to store tasks, but it has been refactored to use an ArrayList instead. This change improves the task display format and allows for dynamic resizing of the task list. The addTask method has been updated to use the add() method of ArrayList to add tasks, and the taskIndex variable has been removed.

matchastack committed:
feat: Add delete functionality for tasks

This commit adds the ability to delete tasks in the ZBot application. The delete functionality allows the user to remove a specific task from the task list by providing the task number as input. If the task number is valid, the task is removed from the list and the updated task count is displayed. If the task number is invalid or not provided, appropriate error messages are shown to the user.

matchastack committed:
chore: Refactor ZBot's main method

This commit refactors the `main` method in the `ZBot` class to improve the handling of user commands. The method now uses `processInput` method to determine the appropriate action based on the user's input. This change simplifies the code and enhances readability.

matchastack committed:
feat: Add Storage class for file handling

This commit adds a new `Storage` class that handles file operations in the ZBot application. The `Storage` class is responsible for creating a file if it does not exist, using the provided file path. This change improves the application's ability to store and retrieve data from files.

matchastack committed:
bug fix: Create tasks.txt file, not folder

This commit fix the `Storage` class to improve file handling in the ZBot application. The `createFileIfNotExists` method previously creates a tasks.txt/ folder instead of tasks.txt file.

matchastack committed:
feat: Save tasks to file

Current program is one-time use and all data is lost upon exit.

This commit adds the feature to save exisiting tasks and their respective details to text file.

matchastack committed:
chore: Move saveTasks method to Storage

The `saveTasks` method a member of Storage class instead of ZBot class. It has been updated to save tasks using an ArrayList.

matchastack committed:
feat: Add loadTasks method

The `loadTasks` method has been added to the Storage class to handle file operations in the ZBot application.
The method load tasks from a file using a comma as a delimiter. This improves the application's ability to save and retrieve tasks from a file.

matchastack committed:
Merge branch 'branch-level-7'

matchastack committed:
chore: Move Storage instance to main method

Initially a static instance under ZBot class due to static method `saveTasks`.

No longer necessary once the `saveTasks` method has been moved to Storage class upon refactoring.

matchastack committed:
refactor: Update Deadline and Event classes to use LocalDateTime

The Deadline and Event classes have been updated to use the LocalDateTime class from the java.time package instead of String for representing dates and times. This change improves the consistency and reliability of date and time handling in the application.

matchastack committed:
Merge branch 'branch-Level-8'

matchastack committed:
feat: Update ZBot to use TaskList instead of ArrayList

The ZBot class `tasks` attribute has been modified to a `TaskList` object instead of an `ArrayList<Task>`. This change improves the code's OOP design and improves maintainability.

matchastack committed:
refactor: Move main logic to `run()`

matchastack committed:
refactor: Add Ui class

A new Ui class has been implemented to handle features pertaining to interactions with the user. This change improves the code's OOP design and improves maintainability.

matchastack committed:
feat: Add print methods to Ui class

The Ui class has been updated with new print methods to display messages for adding, deleting, marking, and unmarking tasks. This improves the user experience by providing clear feedback on task operations.

matchastack committed:
refactor: Rename methods in Storage class

The methods `saveTasks` and `loadTasks` in the Storage class have been renamed to `save` and `load` respectively. This change improves the clarity and consistency of method names in the codebase.

matchastack committed:
refactor: Update Storage.load() to throw ZBotException

The `load()` method in the Storage class has been updated to throw a `ZBotException` when the file is not found. This improves error handling and provides more meaningful feedback to the user.

matchastack committed:
feat: Add printLoadingError() method to Ui class

matchastack committed:
Merge branch 'a-moreOOP'

matchastack committed:
refactor: Move input reading to Ui class

The reading of user input is now handled by the Ui class. The Scanner object in the Ui class is now closed after the outro message is printed. This ensures proper resource management and prevents resource leaks.

matchastack committed:
refactor: Update processInput() method to include Storage parameter

This commit changes the way tasks are saved to file. Previously, tasks are only saved to file on exit. Now, tasks are saved to file on every modification command. This ensures that data can be retrieved even when the program is terminated in an unexpected way.

matchastack committed:
Merge branch 'refactor'

matchastack committed:
refactor: Add zbot and zbot.task package

This commit moves the Parser, Storage, TaskList, Ui, and ZBot classes to the zbot package and Task, ToDo, Deadline, Event classes to the zbot.task package. This improves the organization and structure of the codebase by grouping related classes together in a dedicated package.

matchastack committed:
Merge branch 'add-gradle-support'

matchastack committed:
refactor: Update mainClass to "zbot.ZBot" in build.gradle

Update the mainClass property in the build.gradle file to "zbot.ZBot" to reflect the recent package and class restructuring. This ensures that the correct main class is used when running the application.

matchastack committed:
refactor: Add TaskTest class for unit testing

matchastack committed:
refactor: Add ParserTest class for unit testing

matchastack committed:
refactor: Add StorageTest class for unit testing

matchastack committed:
chore: Update archiveBaseName and archiveFileName in build.gradle

Update the archiveBaseName property in the build.gradle file to "zbot" and the archiveFileName property to "zbot.jar". This ensures that the generated JAR file has the correct name and reflects the recent project renaming.

matchastack committed:
chore: Add Java Doc comments for Task.java

matchastack committed:
chore: Add Java Doc comments for ToDo class

matchastack committed:
chore: Add JavaDoc comments for Deadline class

matchastack committed:
chore: Add JavaDoc comments for Event class

matchastack committed:
chore: Add JavaDoc comments for Parser class

matchastack committed:
chore: Add JavaDoc comments for Storage class

matchastack committed:
chore: Add JavaDoc comments for TaskList class

matchastack committed:
chore: Add JavaDoc comments for Ui class

matchastack committed:
chore: Add JavaDoc comments for ZBot class

matchastack committed:
chore: Refactor ZBotException class

Add blank lines in the ZBotException class for readability.

matchastack committed:
feat: Add findTask() method to ZBot class

This commit adds a new method, `findTask()`, to the `ZBot` class. The method takes in user input and searches for tasks containing a specific keyword. If any tasks are found, they are displayed with their corresponding task numbers. If no tasks are found, a message indicating that no tasks were found is displayed. This new functionality enhances the search capabilities of the application.

matchastack committed:
Merge branch 'branch-Level-9'

matchastack committed:
Merge branch 'branch-A-JavaDoc'

matchastack committed:
refactor: Rename processInput() method to executeInput()

The processInput() method in the ZBot class has been renamed to executeInput() to better reflect its purpose of executing user input. This improves code clarity and maintainability.

matchastack committed:
refactor: Add lines spacing

matchastack committed:
refactor: Add lines spacing for Ui class

matchastack committed:
refactor: Add lines spacing for TaskList class

matchastack committed:
refactor: Add lines spacing for Storage class

matchastack committed:
refactor: Update Storage class to use 'dueDate' instead of 'by'

The Storage class has been updated to use the 'dueDate' variable instead of 'by' for the Deadline task type. This change improves code clarity and consistency.

matchastack committed:
refactor: Update Event class to use 'startDate' and 'endDate' instead of 'from' and 'to'

The Event class has been updated to use the 'startDate' and 'endDate' variables instead of 'from' and 'to' for better clarity and consistency. This change aligns with the recent changes made to the Storage class and improves code maintainability.

matchastack committed:
refactor: Add lines spacing for ToDo class

matchastack committed:
Merge branch 'branch-A-CodingStandard'

matchastack committed:
chore: Add checkstyle plugin and configure version

matchastack committed:
refactor: Fix checkstyleTest violations

matchastack committed:
refactor: Fix checkstyleMain violations

matchastack committed:
Merge branch 'checkstyle'

matchastack committed:
refactor: Update .gitignore to exclude vscode files

matchastack committed:
refactor: Update ZBot class to use DEFAULT_FILE_PATH constant

The ZBot class has been updated to use the DEFAULT_FILE_PATH constant instead of the SAVE_PATH constant for better clarity and consistency. This change aligns with the This change aligns with the addition of a no-argument constructor for ZBot class improves code maintainability.

matchastack committed:
refactor: Update build.gradle to include JavaFX dependencies

The build.gradle file has been updated to include JavaFX dependencies for Windows, Mac, and Linux platforms. This change ensures that the application can run on different operating systems with the required JavaFX libraries.

matchastack committed:
refactor: Add Launcher class to workaround classpath issues

matchastack committed:
feat: Add GUI for ZBot application

This commit introduces a GUI for the existing ZBot program.

This is the first iteration in GUI  development and it does not possess all the features available in the text-based UI in previous commits.

The user is currently able to `list` and add tasks with `todo`, `deadline`, `event`, `mark`, `unmark`, `delete`, `find`. The `bye` command to exit program feature has not been transferred. User should close the GUI window to terminate the program instead.

matchastack committed:
feat: Close the window when user inputs "bye" in handleUserInput() method

df-Nic commented on src/main/java/zbot/Storage.java:
```diff
@@ -0,0 +1,141 @@
+     *
+     * @param text Text to write to the file.
+     */
+    public void writeToFile(String text) {
```
A suggestion will be to indicate the type of file to be written into in the method name.

**For example:**
```java
public void writeToTextFile(String text) {
    // Function Body
}
```

matchastack replied:
makes sense, I will make the change. thanks!

df-Nic commented on src/main/java/zbot/task/Event.java:
```diff
@@ -0,0 +1,50 @@
+    /**
+     * Constructor for Event.
+     *
+     * @param description Description of the event.
```
If I am not mistaken, for JavaDocs there is not need to align the description for each entry?

matchastack replied:
sorry I do not understand what is meant by "align the description for each entry" can you provide some examples? Thank you.

df-Nic replied:
Sorry for the late reply

```
/**
 * Constructor for Event.
 *
 * @param description Description of the event.
 * @param from        Start datetime of the event.
 * @param to          End datetime of the event.
 */
```
From what I see the JavaDoc looks like this.

I just mean that for the other 2 params, `from` and `to` there is no need to have the extra white spaces in-between the variable name and the description of the variable, if I recall correctly.

matchastack replied:
I see what you mean...

Yes, thanks for pointing it out. I agree it's not required but I don't think it violates the standard too as long as it is kept consistent across. 

I will take a look across the other files and choose a format for consistency. Thanks!

df-Nic commented on src/main/java/zbot/ui/MainWindow.java:
```diff
@@ -0,0 +1,58 @@
+import javafx.scene.image.Image;
+import javafx.scene.layout.AnchorPane;
+import javafx.scene.layout.VBox;
+import zbot.ZBot;
```
Possibly add a space between the imports of different packages.

**For example:**
```java
//...
import javafx.scene.layout.VBox;

import zbot.Zbot
```

To follow the standardization of the other imports in the other files, which is good that you have a standard for the other java files as well!

matchastack replied:
yes! I plan to do that but it will not pass the `checkstyle.xml` from AB3. I am still finding out how to edit the file. Thanks for pointing it out though!

df-Nic commented on src/main/java/zbot/Parser.java:
```diff
@@ -0,0 +1,52 @@
+ * Represents a parser that parses and formats date and time.
+ */
+public class Parser {
+    private static final String DEFAULT_TIME = "2359";
```
Good that you follow the snake case naming convention for static variables.

df-Nic commented on src/main/java/zbot/Storage.java:
```diff
@@ -0,0 +1,141 @@
+
+    /**
+     * Creates the file if it does not exist.
+     *
```
A suggestion to add <p> to denote a new paragraph.

If I am not mistaken, this will render the description as 1 line. Therefore possibly try this.

**For example:**
```java
    /**
     * Creates the file if it does not exist.
     * <p>
     * If the file does not exist, the folder containing the file will be created as
     * well.
     * <p>
     * If the file already exists, nothing will be done.
     */
    public void createFileIfNotExists() {
        // Function Body
    }
```

matchastack replied:
I see... did not know that. Thanks!

df-Nic commented on src/main/java/zbot/ZBot.java:
```diff
@@ -0,0 +1,228 @@
+    public String executeInput(String input) {
+        String response;
+
+        if (input.equals("list")) {
```
Good that you follow the indentation format for the `if`, `else` statements 👍

df-Nic reviewed (commented):
Overall the quality of the code follows the standard shown in the lecture materials 💯

matchastack committed:
refactor: Rename writeToFile() method to writeToTextFile() for clarity and consistency

The writeToFile() method in the Storage class has been renamed to writeToTextFile() to improve clarity and consistency in method naming. This change aligns with the recent user commit that introduced a GUI for the ZBot application and ensures that the method name accurately reflects its purpose of writing text to a file.

matchastack committed:
refactor: Improve consistency in print methods

The printAddTaskMsg, printDeleteTaskMsg, printMarkTaskMsg,
and printUnmarkTaskMsg methods in the Ui class have been modified to
improve consistency.

The response messages now start with a new line and end with a new line,
providing a clearer separation between the task details and the additional information.

This change aligns with the recent user commits that introduced a GUI for
the ZBot application and ensures a consistent and user-friendly output format.

matchastack committed:
refactor: Add assertions

Add null checks for task objects in Storage and ZBot classes
to ensure that tasks are not null before performing operations on them.

matchastack committed:
Merge pull request #1 from liauzhanyi/branch-A-Assertions

refactor: Add assertions

matchastack committed:
Enable assertions in run task

matchastack committed:
refactor: Add extra line

The extra line at the end of last method was added
to improve code readability and maintain consistency.

matchastack committed:
refactor: Rename variables

This commit rename some of the variables, attributes and methods
to better represent the value it holds.

matchastack committed:
refactor: Remove comment

matchastack committed:
Refactor Storage class

This commit refactor the Storage class by introducing guard clause
and remove fallthrough else statements.

matchastack committed:
refactor: Improve variable names and code reuse

This commit creates a `modifyTask` method to handle
`markTask`, `unmarkTask` and `deleteTask`.
This helps to reduce repeated code and improves code maintainability.

matchastack committed:
Merge branch 'master' into branch-A-CodeQuality

matchastack committed:
Merge pull request #2 from liauzhanyi/branch-A-CodeQuality

Improve code quality

matchastack committed:
refactor: Rename variable

matchastack committed:
bugfix: Show introduction message in GUI

matchastack committed:
feat: Add functionality to add notes to tasks

- Added a new method `addNoteToTask` in the `ZBot` class
to handle adding notes to tasks.
- Created a new `Note` class to represent a note for a task.
- Modified the `Task` class to include a `Note` object
and added methods to add and retrieve notes.
- Updated the `toString` method in the `Task` class
to display the note if it is not empty.

matchastack committed:
refactor: Abstract add note message to Ui class

matchastack committed:
feat: Store note in file for task with note

- Modified the Storage class to include the note content
in the save file for tasks with notes
- Updated constructors in Task, ToDo, Deadline, and Event classes
to accept note content or note objects
- Added logic to handle the storage and retrieval of note content
in the Storage class

matchastack committed:
Merge tag 'BCD-Extension'

matchastack committed:
bugfix: Fix task numbering logic in ZBot.java

matchastack committed:
chore: Update ZBot User Guide

- Update the title and intro of the user guide to match the product name
- Add a GUI screenshot to the user guide
- Update the quickstart section with instructions
for Java installation and downloading the latest .jar file
- Add examples and descriptions for each command in the features section
- Include a command summary table at the end of the user guide

matchastack committed:
refactor: Update command summary table formatting

matchastack committed:
refactor: Update command summary table formatting

matchastack committed:
docs: Update command summary in README.md

matchastack committed:
refactor: Move Ui.java to ui package

matchastack committed:
chore: Update .jar link

matchastack committed:
Update README.md

matchastack committed:
docs: Add note command to README.md

matchastack committed:
refactor: Rename duke to zbot

matchastack committed:
docs: Update JavaDoc comment

matchastack committed:
feat: Handle trailing whitespace in user input

matchastack committed:
fix: Move file init to Storage constructor

matchastack committed:
docs: Add known issues section to README.md

matchastack committed:
docs: Update known issues section in README.md

matchastack committed:
docs: Update image sources in README.md

matchastack committed:
refactor: Update Storage class delimiter

matchastack committed:
docs: Update README.md
